Sources/PackageGraph/PackageGraphRoot.swift

```diff
@@ -33,23 +33,33 @@ public struct PackageGraphRootInput {
 /// Represents the inputs to the package graph.
 public struct PackageGraphRoot {
 
-    /// The list of root manifests.
-    public let manifests: [Manifest]
+    /// The root packages.
+    public let packages: [PackageIdentity: (manifest: Manifest, packageReference: PackageReference)]
+
+    /// The root manifests.
+    public var manifests: [PackageIdentity: Manifest] {
+        return self.packages.mapValues { $0.manifest }
+    }
 
     /// The root package references.
-    public let packageRefs: [PackageReference]
+    public var packageReferences: [PackageReference] {
+        return self.packages.values.map { $0.packageReference }
+    }
 
     /// The top level dependencies.
     public let dependencies: [PackageDependencyDescription]
 
     /// Create a package graph root.
-    public init(input: PackageGraphRootInput, manifests: [Manifest], explicitProduct: String? = nil) {
-        // TODO: this does not use the identity resolver which is probably fine since its the root packages
-        self.packageRefs = zip(input.packages, manifests).map { (path, manifest) in
-            let identity = PackageIdentity(url: manifest.packageLocation)
-            return .root(identity: identity, path: path)
-        }
-        self.manifests = manifests
+    public init(input: PackageGraphRootInput, manifests: [Manifest], explicitProduct: String? = nil) throws {
+        self.packages = try input.packages.reduce(into: .init(), { partial, inputPath in
+            let manifestPath = inputPath.basename == Manifest.filename ? inputPath : inputPath.appending(component: Manifest.filename)
```

Comment:
Does this need to take into account versioned package manifests?

Reply:
this is mostly in place for testing -- some test pass the full manifest path and some just the directory. I would like to remove this crutch and clean up the tests